complement/pipeline.yml

```diff
@@ -37,3 +40,60 @@ steps:
       automatic:
         - exit_status: 128
           limit: 3
+
+  - command:
+      # Download the latest checkout of Synapse develop
+      - "wget -N https://github.com/matrix-org/synapse/archive/develop.tar.gz -O synapse.tar.gz && tar -xzf synapse.tar.gz && cd synapse"
+      # Build the Synapse production docker image
+      - docker build -t matrixdotorg/synapse -f docker/Dockerfile .
+      # Now hop over to complement's checkout and build the Complement-specific Synapse docker image,
+      # which inherits from the production Synapse Dockerfile
+      - "cd /src && docker build -t complement-synapse -f dockerfiles/Synapse.Dockerfile ."
+      # Run the tests!
+      - "COMPLEMENT_BASE_IMAGE=complement-synapse go test -v -tags "synapse_blacklist,msc2403" ./tests"
+    label: "\U0001F9EA Complement / Synapse Monolith / :go: 1.15"
+    agents:
+      queue: "medium"
+    plugins:
+      - docker#v3.7.0:
+          # The dockerfile for this image is at https://github.com/matrix-org/complement/blob/master/dockerfiles/ComplementCIBuildkite.Dockerfile.
+          image: "matrixdotorg/complement:latest"
+          mount-buildkite-agent: false
+          # Complement needs to know if it is running under CI
+          environment:
+           - "CI=true"
+          publish: [ "8448:8448" ]
+          # Complement uses Docker so pass through the docker socket. This means Complement shares
+          # the hosts Docker.
+          volumes:
+            - "/var/run/docker.sock:/var/run/docker.sock"
+    retry:
+      automatic:
+        - exit_status: 128
+          limit: 3
+
+  - command:
+      # Build the Dendrite for Complement docker image
+      - "cd /src && docker build -t complement-dendrite -f dockerfiles/Dendrite.Dockerfile ."
+      # Run the tests!
+      - "COMPLEMENT_BASE_IMAGE=complement-dendrite go test -v -tags "msc2836" ./tests"
+    label: "\U0001F9EA Complement / Dendrite Monolith / :go: 1.15"
+    agents:
+      queue: "medium"
+    plugins:
+      - docker#v3.7.0:
+          # The dockerfile for this image is at https://github.com/matrix-org/complement/blob/master/dockerfiles/ComplementCIBuildkite.Dockerfile.
+          image: "matrixdotorg/complement:latest"
+          mount-buildkite-agent: false
+          # Complement needs to know if it is running under CI
+          environment:
+           - "CI=true"
+          publish: [ "8448:8448" ]
```

Comment:
Have you tried running this pipeline locally? I doubt this will work because synapse/dendrite will be fighting for the same ports.

Reply:
I did not, and thank you for the suggestion, I discovered a few bugs along the way! I also wrote some documentation on running pipelines locally: https://github.com/matrix-org/pipelines/pull/116 If you often do so yourself, can you take a look and see if what I was doing is overkill?

Otherwise I didn't experience any issues with port allocation during my test runs. Both Synapse and Dendrite are created in separate build steps, so they shouldn't interfere with one another?

Reply:
it'll presumably fail if you try to run both buildsteps on the same machine, but none of our build agents are set up to do that, so you can probably get away with it.

Reply:
I ran the entire pipeline solely on my machine though.

Reply:
sorry, what I mean is if you run both build steps *concurrently* on one machine, either by configuring the buildkite agent process to run multiple build steps at once, or by running multiple buildkite agent processes. 

Presumably you did neither of the above, but what I'm saying is, that's fine. None of the buildkite agents we use in our "production" buildkite do so either.

Reply:
Presumably this concern may be further eliminated by now using burn-after-build agents.